ddtrace/tracer.py

```diff
@@ -890,6 +891,60 @@ def flush(self):
         """Flush the buffer of the trace writer. This does nothing if an unbuffered trace writer is used."""
         self._writer.flush_queue()
 
+    def wrap_generator(
+        self,
+        name=None,  # type: Optional[str]
+        service=None,  # type: Optional[str]
+        resource=None,  # type: Optional[str]
+        span_type=None,  # type: Optional[str]
+    ):
+        def wrap_decorator(f):
+            # type: (AnyCallable) -> AnyCallable
+            # FIXME[matt] include the class name for methods.
+            span_name = name if name else "%s.%s" % (f.__module__, f.__name__)
+
+            # detect if the the given function is a coroutine to use the
+            # right decorator; this initial check ensures that the
+            # evaluation is done only once for each @tracer.wrap
+            if compat.iscoroutinefunction(f):
+                # call the async factory that creates a tracing decorator capable
+                # to await the coroutine execution before finishing the span. This
+                # code is used for compatibility reasons to prevent Syntax errors
+                # in Python 2
+                func_wrapper = compat.make_async_decorator(
+                    self,
+                    f,
+                    span_name,
+                    service=service,
+                    resource=resource,
+                    span_type=span_type,
+                )
+            else:
+
+                @functools.wraps(f)
+                def func_wrapper(*args, **kwargs):
+                    # if a wrap executor has been configured, it is used instead
+                    # of the default tracing function
+                    if getattr(self, "_wrap_executor", None):
+                        yield from self._wrap_executor(
```

Comment:
I tried looking into what this is and I _think_ this is correct to still use `yield from`. But let me know if I am wrong :) I could also add a test case if you point me to an appropriate one to base one off

Reply:
Ah, right - this doesn't work on Python 2. It's probably the root cause of the [test failures](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/31969/workflows/e0e7d56f-5641-40e5-9db1-e059f9e5a40e/jobs/2163071):
```
Traceback (most recent call last):
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/bin/pytest", line 8, in <module>
    sys.exit(main())
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/_pytest/config/__init__.py", line 65, in main
    config = _prepareconfig(args, plugins)
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/_pytest/config/__init__.py", line 214, in _prepareconfig
    pluginmanager=pluginmanager, args=args
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/pluggy/manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/pluggy/callers.py", line 203, in _multicall
    gen.send(outcome)
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/_pytest/helpconfig.py", line 94, in pytest_cmdline_parse
    config = outcome.get_result()
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/pluggy/callers.py", line 81, in get_result
    _reraise(*ex)  # noqa
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/_pytest/config/__init__.py", line 789, in pytest_cmdline_parse
    self.parse(args)
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/_pytest/config/__init__.py", line 997, in parse
    self._preparse(args, addopts=addopts)
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/_pytest/config/__init__.py", line 943, in _preparse
    self.pluginmanager.load_setuptools_entrypoints("pytest11")
  File "/root/project/.riot/venv_py2718_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis6451_boto_moto100/lib/python2.7/site-packages/pluggy/manager.py", line 299, in load_setuptools_entrypoints
    plugin = ep.load()
  File "/root/project/.riot/venv_py2718/lib/python2.7/site-packages/importlib_metadata/__init__.py", line 105, in load
    module = import_module(match.group('module'))
  File "/root/.pyenv/versions/2.7.18/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/root/project/ddtrace/__init__.py", line 17, in <module>
    from .tracer import Tracer  # noqa: E402
  File "/root/project/ddtrace/internal/module.py", line 186, in _load_module
    module = self.loader.load_module(fullname)
  File "/root/.pyenv/versions/2.7.18/lib/python2.7/pkgutil.py", line 243, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "/root/project/ddtrace/tracer.py", line 928
    yield from self._wrap_executor(
             ^
SyntaxError: invalid syntax
```
You might try replacing `yield from` with [older syntax](https://stackoverflow.com/a/17581397).